File: lib/py/telepact/internal/binary/ServerBase64Encode.py
#|
#|  Copyright The Telepact Authors
#|  SPDX-License-Identifier: Apache-2.0
#|

from typing import TYPE_CHECKING, cast
import base64

def server_base64_encode(message: list[object]) -> None:
    headers = cast(dict[str, object], message[0])
    body = cast(dict[str, object], message[1])

    base64_paths = cast(dict[str, object], headers.get("@base64_", {}))

    travel_base64_encode(body, base64_paths)

def travel_base64_encode(value: object, base64_paths: object) -> object:
    if isinstance(base64_paths, dict):
        for key, val in base64_paths.items():
            if val is True:
                if key == "*" and isinstance(value, list):
                    for i, v in enumerate(value):
                        nv = travel_base64_encode(v, val)
                        value[i] = nv
                elif key == "*" and isinstance(value, dict):
                    for k, v in value.items():
                        nv = travel_base64_encode(v, val)
                        value[k] = nv
                elif isinstance(value, dict):
                    nv = travel_base64_encode(value[key], val)
                    value[key] = nv
                else:
                    raise ValueError(f"Invalid base64 path: {key} for value: {value}")
            else:
                if key == "*" and isinstance(value, list):
                    for i, v in enumerate(value):
                        travel_base64_encode(v, val)
                elif key == "*" and isinstance(value, dict):
                    for k, v in value.items():
                        travel_base64_encode(v, val)
                elif isinstance(value, dict):
                    travel_base64_encode(value[key], val)
                else:
                    raise ValueError(f"Invalid base64 path: {key} for value: {value}")
        return None
    elif base64_paths is True and (isinstance(value, bytes) or value is None):
        if value is None:
            return None
        return base64.b64encode(value).decode("utf-8")
    else:
        raise ValueError(f"Invalid base64 path: {base64_paths} for value: {value}")

File: lib/ts/src/internal/binary/ServerBase64Encode.ts
import { Buffer } from "buffer";

export function serverBase64Encode(message: object[]): void {
    const headers = message[0] as Record<string, any>;
    const body = message[1] as Record<string, any>;

    const base64Paths = headers["@base64_"] as Record<string, any> || {};

    travelBase64Encode(body, base64Paths);
}

function travelBase64Encode(value: any, base64Paths: any): any {
    if (typeof base64Paths === "object" && !Array.isArray(base64Paths)) {
        for (const [key, val] of Object.entries(base64Paths)) {
            if (val === true) {
                if (key === "*" && Array.isArray(value)) {
                    for (let i = 0; i < value.length; i++) {
                        value[i] = travelBase64Encode(value[i], val);
                    }
                } else if (key === "*" && typeof value === "object" && value !== null) {
                    for (const k in value) {
                        value[k] = travelBase64Encode(value[k], val);
                    }
                } else if (typeof value === "object" && value !== null) {
                    value[key] = travelBase64Encode(value[key], val);
                } else {
                    throw new Error(`Invalid base64 path: ${key} for value: ${value}`);
                }
            } else {
                if (key === "*" && Array.isArray(value)) {
                    for (const v of value) {
                        travelBase64Encode(v, val);
                    }
                } else if (key === "*" && typeof value === "object" && value !== null) {
                    for (const v of Object.values(value)) {
                        travelBase64Encode(v, val);
                    }
                } else if (typeof value === "object" && value !== null) {
                    travelBase64Encode(value[key], val);
                } else {
                    throw new Error(`Invalid base64 path: ${key} for value: ${value}`);
                }
            }
        }
        return null;
    } else if (base64Paths === true && (Buffer.isBuffer(value) || value === null)) {
        if (value === null) {
            return null;
        }
        return Buffer.from(value).toString("base64");
    } else {
        throw new Error(`Invalid base64 path: ${base64Paths} for value: ${value}`);
    }
}